dotnet/src/InternalUtilities/src/Diagnostics/ModelDiagnostics.cs

```diff
@@ -0,0 +1,244 @@
+        TryAddTag("top_p", ModelDiagnosticsTags.TopP);
+    }
+
+    /// <summary>
+    /// Convert chat history to a string aligned with the OpenAI format
+    /// </summary>
+    private static string ToOpenAIFormat(IEnumerable<ChatMessageContent> chatHistory)
+    {
+        return $"[{string.Join(
+            ", \n",
+            chatHistory.Select(m => $"{{\"role:\" {m.Role}, \"content:\" {m.Content}}}"))}]";
+    }
+
+    /// <summary>
+    /// Start a completion activity and return the activity.
+    /// The `formatPrompt` delegate won't be invoked if events are disabled.
+    /// </summary>
+    private static Activity? StartCompletionActivity<T>(
+        Uri? endpoint,
+        string modelName,
+        string modelProvider,
+        T prompt,
+        PromptExecutionSettings? executionSettings,
+        Func<T, string> formatPrompt)
+    {
+        if (!IsModelDiagnosticsEnabled())
+        {
+            return null;
+        }
+
+        string operationName = prompt is ChatHistory ? "chat.completions" : "text.completions";
+        var activity = s_activitySource.StartActivityWithTags(
+            $"{operationName} {modelName}",
+            [
+                new(ModelDiagnosticsTags.Operation, operationName),
+                new(ModelDiagnosticsTags.System, modelProvider),
+                new(ModelDiagnosticsTags.Model, modelName),
+            ]);
+
+        if (endpoint is not null)
+        {
+            activity?.AddTags([
+                new(ModelDiagnosticsTags.Address, endpoint.AbsoluteUri),
+                new(ModelDiagnosticsTags.Port, endpoint.Port),
+            ]);
+        }
+
+        AddOptionalTags(activity, executionSettings);
+
+        if (s_enableSensitiveEvents)
+        {
+            var formattedContent = formatPrompt(prompt);
+            activity?.AttachSensitiveDataAsEvent(
+                ModelDiagnosticsTags.PromptEvent,
+                [
+                    new(ModelDiagnosticsTags.PromptEventPrompt, formattedContent),
+                ]);
+        }
+
+        return activity;
+    }
+
+    /// <summary>
+    /// Set the completion response for a given activity.
+    /// The `formatCompletions` delegate won't be invoked if events are disabled.
+    /// </summary>
+    private static void SetCompletionResponse<T>(
+        Activity activity,
+        T completions,
+        int? promptTokens,
+        int? completionTokens,
+        Func<T, string> formatCompletions) where T : IEnumerable<KernelContent>
+    {
+        if (!IsModelDiagnosticsEnabled())
+        {
+            return;
+        }
+
+        activity.AddTags(
+            [
+                new(ModelDiagnosticsTags.FinishReason, GetFinishReasons(completions)),
+                new(ModelDiagnosticsTags.PromptToken, promptTokens ?? -1),
+                new(ModelDiagnosticsTags.CompletionToken, completionTokens ?? -1),
+                new(ModelDiagnosticsTags.ResponseId, GetResponseId(completions.FirstOrDefault())),
```

Comment:
Should we keep the tags with some default values if they are not available or we don't include the tags when the values aren't available?

Reply:
we should not include them at all

Reply:
Tags should be dropped if the values are not available.